added implementation of DRY sampler (post-refactor)

wwoodsTM opened the pull request:
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [x] Medium
  - [ ] High

This is a continuation of the previous PR #6839 to implement DRY sampling (and I think a pretty faithful port of Koboldcpp’s impressive implementation, authored by @pi6am [who is also credited as co-author for this PR], which in turn is originally based on @p-e-w‘s design and ideas as explained [here](https://github.com/oobabooga/text-generation-webui/pull/5677)), but now overhauled to use the new sampler API. I added @l3utterfly as co-author in a recent commit of this PR.

Here are some important notes (**EDIT**: all have been addressed at this point, and in the case of the first point, this was moved to a separate PR, #10019):
- ~I may have also fixed #8971 (or it may turn out to be more of a bandaid), but by moving/forcing sampling to happen over the entire prompt at the conclusion of prompt processing in server.cpp, I was able to get consistent sampling of the prompt, regardless of caching or whether it is the first generation or not.~
- ~I also fixed a recent breaking change that was made, though I am not sure when this change happened. For quite a while it was possible (and indeed I have heavily relied on the feature in my own software for a long while now) to be able to submit a mixed array of strings and token IDs as a singleton prompt. This I believe was originally implemented from PR #2306. Apologies if this is considered out of scope here, and I can remove it or move it to a separate PR if need be, but to be honest this feature was the main reason why I converted to using llama.cpp's server. It also seems this fix restores the feature and doesn't stop the other prompting options from working either, but if I missed something here, please let me know.~
- ~I added a context size parameter to gpt_sampler_init(). If there is a smarter way to handle this, I am happy to do it in a less drastic way, but I couldn't immediately figure that out.~
- ~The classic issue that I have dealt with from day one with this where I can't figure out how to get the vector of strings for the sequence breakers from server.cpp/common/API land to src/llama-sampling land is still here. I again "cheated" with a C++ section at the bottom of llama.h and made a separate function for setting the sequence breakers apart from the regular init function. There is probably some clever way to avoid this or to use a void or extern that I am not as experienced with. I welcome any help or suggestions with that or any of these other issues.~
- ~Finally, I did not implement a CLI argument for sequence breakers. If anybody wants to take a whack at that, please feel free. My thought was that maybe it could be done similarly to how the reverse/anti prompt is done (otherwise known as stopping strings) or how JSON schema is done in the CLI, but I wasn't really sure how to pull it off correctly.~

wwoodsTM commented:
@compilade You had kindly offered possibly doing a code review for the previous iteration of this PR before the sampler API overhaul. If you still have availability for that, I would definitely love to get your feedback and insights on this newest iteration.

wwoodsTM commented on examples/server/server.cpp:
```diff
@@ -1476,20 +1504,32 @@ struct server_context {
+            return has_number;
+        };
+
+        bool is_singleton = prompt.is_string() || (prompt.is_array() && is_valid_singleton_array(prompt));
```
Just realized I accidentally left this "is_singleton" variable in here when I had meant to take it out. I will remove it in my next commit.

wwoodsTM commented:
@compilade Thanks for taking a look! If it’s any help the [koboldcpp version](https://github.com/LostRuins/koboldcpp/blob/c38d1ecc8d8f4bcfe2f959584002ddf7a80b8e9b/gpttype_adapter.cpp#L743) of this has a lot of extra comments that go more in-depth on the intricacies of the Z-algorithm and the clever way they implement it. Most of the core implementation here is exactly the same except for the bits I had to change due to the ring_buffer accessing from the end.

p-e-w commented:
KoboldCPP's handling of sequence breakers is excellent, and it's great to see it ported here.

However, the Z-algorithm is actually unnecessary complexity as I had remarked in the Kobold PR. Quadratic runtime can be avoided by simply capping the maximum sequence length being matched (to 50, in the case of text-generation-webui). The penalty is exponential in the match length, so this makes no difference in practice as the value will be astronomically large regardless. This can simplify the algorithm considerably, and has been tested to perform well even in the worst-case scenario where the input consists entirely of repetitions of a single token.

Personally, I prefer simplicity over notions of algorithmic elegance if performance is demonstrably equivalent as in this case, although I accept and respect that not everyone shares that opinion.

Either way, great to see this happening! Make sure you run some basic smoke tests before merging this in, e.g. using the input from the "Demonstration" section of my original PR, to verify the system is actually doing what it is supposed to, since the implementation is fairly complex at this point.

wwoodsTM commented:
> Personally, I prefer simplicity over notions of algorithmic elegance if performance is demonstrably equivalent as in this case, although I accept and respect that not everyone shares that opinion.

@p-e-w I really appreciate your reply, and am a fan of not just this sampler but also XTC, which I was happy to see at least two people beating me to the punch on as far as implementing with this new sampler API (albeit one of the submissions being made through reddit rather than a PR).

As for the issues you raise about complexity, you bring up very fair points. I will admit that I do tend to be a sucker for "elegant" algorithms or ones that seem particularly clever to me. Having said that, I would not have been opposed to implementing it the less complex way but didn't realize you weren't a fan of the z-algorithm portion until I was already pretty far into the process and saw your comments while digging deeper into all the PRs.

> Either way, great to see this happening! Make sure you run some basic smoke tests before merging this in, e.g. using the input from the "Demonstration" section of my original PR, to verify the system is actually doing what it is supposed to, since the implementation is fairly complex at this point.

In regard to your demonstration text, that was indeed one of the first things I tested it on! For anybody else looking at this, here is the first part of it that is fed to the LLM (in this case, I fed it to Nemo):

```
Detective: Where were you last night at 6 PM?

Suspect: On the advice of my attorneys, I invoke my Fifth Amendment right to not answer that question.

Detective: Did you know the victim personally?

Suspect: On the advice of my attorneys, I invoke my Fifth Amendment right to not answer that question.

Detective: Do you have money problems?

Suspect: On the advice of my attorneys, I invoke my Fifth Amendment right to not answer that question.

Detective: Do you have a criminal record?

Suspect: On the advice of my attorneys, I invoke
```
Here is what Nemo gave me with DRY off when I ran it just now: 
```
 my Fifth Amendment right to not answer that question.

Detective: What did you do for a living?

Suspect: On the advice of my attorneys, I invoke my Fifth Amendment right to not answer that question.

Detective: Have you ever been arrested?

Suspect: On the advice of my attorneys, I invoke my Fifth Amendment right to not answer that question.

Detective: Are you married?

Suspect: On the advice of my attorneys, I invoke my Fifth Amendment right to not answer that question.

...plus 12 more successive repetitions of this pattern as I had banned EOS...
```
Here is what Nemo gave me with this implementation of DRY turned on (multiplier: 0.8, base: 1.75, allowed_length: 2, default sequence breakers which include ":"):
```
 —

Detective: Wait, you're not going to invoke your Fifth Amendment rights on that question?

Suspect: No. I don't have a criminal record.

...plus many more lines, none having that pattern...
```

Here is some relevant debug output showing the final tokens of the prompt, as well as the penalty itself and a comparison of logits before and after:
```
109 | 72859  | Sus        
110 | 2117   | pect       
111 | 1058   | :          
112 | 3249   |  On        
113 | 1278   |  the       
114 | 19199  |  advice    
115 | 1307   |  of        
116 | 2036   |  my        
117 | 101642 |  attorneys 
118 | 1044   | ,          
119 | 1362   |  I         
120 | 45589  |  invoke    

Applied penalty 40.2121 to token 2036 ( my) (repeat length 9)

Top 10 candidates before penalties:
Token 2036 ( my): logit = 23.2354
Token 1674 (—): logit = 14.6016
Token 9283 (...\n\n): logit = 14.5702
Token 1278 ( the): logit = 14.3120
Token 38659 ( ...\n\n): logit = 13.9730
Token 2251 ( —): logit = 13.8171
Token 2880 (...): logit = 12.9741
Token 57725 ( Fifth): logit = 12.7038
Token 2247 ( .): logit = 12.5666
Token 10260 (…\n\n): logit = 12.5319

Top 10 candidates after penalties:
Token 1674 (—): logit = 14.6016
Token 9283 (...\n\n): logit = 14.5702
Token 1278 ( the): logit = 14.3120
Token 38659 ( ...\n\n): logit = 13.9730
Token 2251 ( —): logit = 13.8171
Token 2880 (...): logit = 12.9741
Token 57725 ( Fifth): logit = 12.7038
Token 2247 ( .): logit = 12.5666
Token 10260 (…\n\n): logit = 12.5319
Token 1536 ( by): logit = 12.3046
```

p-e-w commented:
@wwoodsTM 

Fair enough, as long as everything works it should be fine, and since you've already implemented it I guess there's little point in changing it now. The elephant in the room with such decisions is usually maintainability, but to be frank I doubt this code will need to be touched very often.

BTW, I love the output you got with the demonstration example. It shows not only that DRY is capable of stopping repetition (which is a mathematical certainty in principle), but that the model recovers *organically* from the intervention, creating a natural transition to a completely different train of thought. Many people have told me that they expect DRY to just make the model paraphrase the previous output instead of repeating it verbatim, or introduce spelling mistakes, but your output demonstrates that this usually doesn't happen with current-gen models. And it's even less likely to happen in actual conversations, because in such scenarios, DRY gradually increases the penalty rather than just slamming the door because the user already imposed a 9-token repetition like in this example.

wwoodsTM commented:
@compilade I went ahead and made several small fixes that hopefully should address the strict compile warnings/errors I saw from the workflows and also removed the redundant line I had commented on earlier. Thanks again and let me know if I can do anything else to help the process.

slaren commented:
There is a lot of code here that makes this change hard to review. It would be easier to do so if all the debug code was removed. Functions like `llama_tokenize` in `llama-sampling.cpp` should probably be removed as well, or moved elsewhere.

wwoodsTM commented:
@slaren Thank you for the feedback!

- [x] I just removed all of the debugging code that had been between the `#ifdefs`.
- [x] I also moved the tokenize/detokenize functions to `llama-impl.h`

Despite it not being a strict interface function, I left the `GetOverlappingTokenSequences` in `llama-sampling.cpp` since it's an integral part to processing the sequence breakers. If you think I should also move that out or rename it to make the connection to DRY more obvious though, I can do that too.

About an hour ago I resolved the recent conflicts that arose from #9805. Prior to that, all of the CI workflows that @compilade triggered were showing as passing.

compilade commented on src/llama-sampling.cpp:
```diff
@@ -1565,6 +1565,292 @@ struct llama_sampler * llama_sampler_init_penalties(
+        return;
+    }
+
+    // Step 2: Z-algorithm implementation
```
> If it’s any help the [koboldcpp version](https://github.com/LostRuins/koboldcpp/blob/c38d1ecc8d8f4bcfe2f959584002ddf7a80b8e9b/gpttype_adapter.cpp#L743) of this has a lot of extra comments that go more in-depth on the intricacies of the Z-algorithm and the clever way they implement it. Most of the core implementation here is exactly the same except for the bits I had to change due to the ring_buffer accessing from the end.

Oh yes this is very useful. Then it might be relevant to link to where the implementation comes from in a comment near the code. Not sure if `kobold.cpp` being under AGPLv3 will cause problems, though.

(cc @pi6am because they wrote <https://github.com/LostRuins/koboldcpp/pull/982>)

wwoodsTM replied:
Oh… yikes. 😬

I feel very naive now for not looking deeper into the licensing ramifications there. My apologies.

I doubt that @LostRuins would want to relicense just this one portion of his code, which is of course totally understandable, so my assumption is barring that, this very likely forces out this critical part of the PR. 

While I really do admire @pi6am’s implementation, I think the best move at this point is to try returning to @l3utterfly’s and my older version of the code as a base (so no Z-algorithm, which I know @p-e-w prefers to not use anyway, so there’s that at least!). 

If I can get that plugged in and working well I will try to rewrite the overlapping sequence breaker code myself so that we are in the clear from a licensing perspective.

Anyway, thanks for all the feedback @compilade, I will focus on this issue first since it obviously impacts such a critical part of the code.

p-e-w replied:
The copyright holder of that portion of Kobold's code is the author of the implementation, @pi6am. Not the maintainer of the project (unless the two signed a copyright assignment agreement). So you would need their permission to relicense that code, which they might be willing to give.

Personally, I think that the MIT license used by llama.cpp is a terrible choice for projects like this, and this issue is one of the reasons. Licensing under MIT or other permissive licenses basically means you're gifting your code to businesses to put behind closed gates, while at the same time limiting what code you yourself can reuse from other open source projects. It's the worst of both worlds.

wwoodsTM replied:
@p-e-w Well, that is fascinating, thank you for clarifying that. I am learning a lot through this whole process!

So to follow up on this idea, I went ahead and just posted a message on the original Koboldcpp PR page pinging @pi6am in the hopes that a notification through there might work better than one here.

LostRuins replied:
Yup you'd have to get permission from @pi6am as they are the author. I personally have no issues.

hgftrdw45ud67is8o89 replied:
yea MIT is a bad choice in this cruel modern world. 
Unless we assume LM studio or x other million profit business actually alllocate 10% of their annual budget back to foss projects that they used. But it's all too late unless ggerna bro and other mantainer decide to change it.

hgftrdw45ud67is8o89 replied:
But assuming that even if we get both algorithm to work,you would still need a proper
-cli flag 
-documentation(or else no one know there were 2 algorithm actually implemented.lol. Precious effort undiscovered by thousands of people)
-frontend that properly implement to utilize it like(sillytavern implement it,etc,or else someone that use frontend may not have it enabled,UNLESS llama.cpp can *force* using a certain algorithm without needing any input from frontend)

All complete to fully benefit everyone.

pi6am replied:
I don't have any objections to porting any or all of my DRY implementation from koboldcpp into llama.cpp. I officially give permission to incorporate it into llama.cpp under the MIT license.

That said, it would be nice to credit the koboldcpp implementation in comments. I also think including some of the other comments from my implementation would be worthwhile to make the code easier to understand.

In particular, I didn't write the Z-algorithm implementation from scratch; I ported a public-domain Python implementation from https://github.com/ivanyu/string-algorithms/blob/master/z_algorithm.py which I credited in a comment. It's not necessary to provide credit to this based on its license, but I think it's polite to do so.

wwoodsTM replied:
> I don't have any objections to porting any or all of my DRY implementation from koboldcpp into llama.cpp. I officially give permission to incorporate it into llama.cpp under the MIT license.

That is kind of you and very appreciated @pi6am, thank you! I had come up with a workable solution that implemented multi-token sequence breakers mixed with the non-z-algorithm approach that I was going to try to implement if your code was not available, but I will be the first to admit that it was nowhere near as optimized as your code, so I really appreciate having your blessing and being able to use your implementation.

> That said, it would be nice to credit the koboldcpp implementation in comments. I also think including some of the other comments from my implementation would be worthwhile to make the code easier to understand.
> 
> In particular, I didn't write the Z-algorithm implementation from scratch; I ported a public-domain Python implementation from https://github.com/ivanyu/string-algorithms/blob/master/z_algorithm.py which I credited in a comment. It's not necessary to provide credit to this based on its license, but I think it's polite to do so.

Agreed, and done on all counts in the commit I just made. Thanks again!

compilade reviewed (commented):
> It looks like @compilade has been showing a busy status since the weekend and I’m guessing has a pretty full plate at the moment.

@wwoodsTM Correct, I've had much less free time that I'd like in the last weeks. But I still want to give you feedback.

I really appreciate the efforts you've put into documenting your changes in the PR description.

wwoodsTM commented on examples/server/server.cpp:
```diff
@@ -885,30 +885,54 @@ struct server_context {
+                    if (dry_sequence_breakers->is_array()) {
+                        slot.sparams.dry_sequence_breakers = dry_sequence_breakers->get<std::vector<std::string>>();
+                    } else if (dry_sequence_breakers->is_string()) {
+                        slot.sparams.dry_sequence_breakers = json::parse(dry_sequence_breakers->get<std::string>()).get<std::vector<std::string>>();
```
I hear you, as it does seem unconventional. My understanding though is that this follows from the implementations in TextGen WebUI, Koboldcpp and SillyTavern, so my aim was for consistenty and interoperability between these implementations and to match people's expectations.

slaren replied:
I think this should be reviewed. To me, it seems very wrong to insert JSON strings within JSON, and other projects that we don't aim to be compatible with doing this is not a good enough reason to keep it this way. @ngxson @ggerganov please take a look.

ngxson replied:
I would say that I'm not a fan of this patch either, looks like the result of a miscommunication between back/front developers.

What I'm thinking is, maybe we can just return an error with the case this is not an array: `Error: dry_sequence_breakers must be a non-empty array of strings`. If enough users want to have this compat, we will add it later.

We should also leave a comment in this code to mark it as a patchy thing. Something like:

```
// Currently, this is not compatible with TextGen WebUI, Koboldcpp and SillyTavern format
// Ref: https://github.com/oobabooga/text-generation-webui/blob/d1af7a41ade7bd3c3a463bfa640725edb818ebaf/extensions/openai/typing.py#L39
```

wwoodsTM replied:
I have gone ahead and implemented your suggestion in this latest commit. @ngxson.

wwoodsTM commented:
We now have the explicit permission from @pi6am to base this version on his implementation from koboldcpp. I also added him as a co-author on earlier commits for this PR.

I believe I had addressed all of the concerns from @compilade's code review in my latest commit, but I of course welcome any and all additional feedback. @ggerganov @slaren 

Thanks.

slaren commented on src/llama-sampling.cpp:
```diff
@@ -1660,6 +1660,446 @@ struct llama_sampler * llama_sampler_init_penalties(
+};
+
+// Ported from Koboldcpp, original PR: https://github.com/LostRuins/koboldcpp/pull/982 (Original author: pi6am)
+static void GetOverlappingTokenSequences(const struct llama_model * model, const std::string& str, std::unordered_multimap<llama_token, std::vector<llama_token>>& token_sequences, int max_tail_len = -1) {
```
Use the same style as everywhere else, eg.:
```suggestion
static void get_overlapping_token_sequences(const struct llama_model * model, const std::string & str, std::unordered_multimap<llama_token, std::vector<llama_token>> & token_sequences, int max_tail_len = -1) {
```

slaren commented on src/llama-sampling.cpp:
```diff
@@ -1660,6 +1660,446 @@ struct llama_sampler * llama_sampler_init_penalties(
+    };
+}
+
+void llama_sampler_dry_set_seq_breakers(struct llama_sampler * smpl, const std::vector<std::string>& seq_breakers) {
```
What is the reason for having a separate function to set this, instead of doing it during initialization in `llama_sampler_init_dry`, like other sampler parameters?

wwoodsTM replied:
In one of @compilade’s previous comments that I marked as resolved above we were discussing the best way to use a vector of C++ strings for the sequence breakers while also allowing for there to be a C interface version of it, so this was one of the possible compromises, having a unified main init function and then a llama.h C version of the set set_seq_breakers function that takes in C-strings and a standard internal version that takes in C++ vectors of std::string’s.

slaren replied:
The same applies to `llama_sampler_dry_set_seq_breakers_c` (which is the function I was thinking about when I wrote that). Generally the design of the samplers is that their parameters are immutable once created, and I don't see a good reason to make an exception here. As far as I can tell, this function is not exposed in the API anywhere (and to be clear, it shouldn't, the llama.cpp API needs to be strictly compatible with C), so it should be made static.

slaren commented on src/llama-sampling.cpp:
```diff
@@ -1660,6 +1660,446 @@ struct llama_sampler * llama_sampler_init_penalties(
+    cur_p->sorted = false;
+}
+
+static void llama_sampler_dry_reset(struct llama_sampler * smpl) {
+    auto * ctx = (llama_sampler_dry *) smpl->ctx;
+    ctx->last_tokens.clear();
+    ctx->dry_repeat_count.clear();
+    ctx->dry_max_token_repeat.clear();
+}
```
Unless `dry_processed_breakers` is moved to the init function, it should also be cleared here to reset the sampler to its original state.

slaren commented on src/llama-sampling.cpp:
```diff
@@ -1660,6 +1660,446 @@ struct llama_sampler * llama_sampler_init_penalties(
+        }
+    }
+}
+
+
+
```
Remove extra lines.

wwoodsTM commented:
@slaren I implemented your suggestions and fixed a conflict with the latest version of master.

Continuing the conversation on the sequence breaker functions: While I had only meant these to be a "part 2" of the init process, I can see how this was not clear and how regardless of intention they don't mesh well with the design.

To address this, I came up with a way of getting rid of the sequence breaker functions altogether which I just implemented in the latest commit. This new approach uses one main init function which include the sequence breakers as a parameter, and then I made two overloaded wrapper versions, one for the C-interface and another for test-sampling.cpp.

Does this approach seem workable?

slaren commented:
Why no take `char**` instead and make it part of the public API?

wwoodsTM commented:
@slaren 

> Why no take `char**` instead and make it part of the public API?

So as @compilade mentioned before, I preferred the flexibility of C++ strings to allow for the possibility of NUL characters as I wasn't sure whether there might be some contexts where this literal "sequence breaker" for C-strings might be used as a sequence breaker for DRY. I also wasn't sure whether there might be any other weird encoding things that might crop up with C-strings. But maybe the chances are quite remote of this?

Also, in my particular use case, I use llama-server and pass sequence breakers through JSON, and so this probably biases me a bit toward thinking of that as kind of the principal interface versus the C-interface.

As a side note, I just made an additional commit to address warnings-made-errors from the pedantic runs that I had missed.

slaren commented:
I am not familiar with this sampler, and I am not sure what kinds of strings are used as sequence breakers, but it seems very unlikely that allowing NUL characters is important. If it is really necessary, it can be done by passing an additional array with the sizes of each string.

wwoodsTM commented:
@slaren 

Do you feel this is a deal-breaker in its current form? My feeling is that at present it is a decent compromise to allow for keeping things in pure C++ string form where possible, such as when the breakers are submitted through JSON and tokenized in get_overlapping_token_sequences, as opposed to converting to C strings and then converting back to C++ strings every time. In the current implementation, sequence breakers would only be converted from C-strings on an as-needed basis when they are specifically submitted through the C-interface.

slaren commented:
My opinion is that it is not worth the maintenance cost of having "secret" API functions, especially for something as minor as this. It takes two lines of code to convert a vector of `std::string` to a vector of `const char *`, and it only needs to be done once in `sampling.cpp`.

wwoodsTM commented:
I hear you, it’s just that C-strings have a history of sending shivers down my spine and making me curse in repetitive ways that would probably be penalized by this sampler. I was hoping to avoid converting the overlap code from kobold to use them, but honestly it shouldn’t be hard since we’re not really manipulating the strings or doing anything complicated with them, so I will get over my fear of them and make the change a bit later today when I get home.

Thanks again for your feedback.

slaren commented:
I am not sure if I understand correctly, but it is ok to store the strings in the sampler state as `std::string`, there is no cost to doing that, since you need to make a copy of the strings anyway to store them (can't just steal the pointers from the application).

slaren commented on src/llama-impl.h:
```diff
@@ -70,6 +70,44 @@ const std::vector<std::pair<std::string, struct ggml_tensor *>> & llama_internal
     struct llama_context * ctx
 );
 
+static std::vector<llama_token> llama_tokenize(
```
I think this function could be replaced by calling `llama_tokenize_internal` instead, which is already exposed internally in `llama-vocab.h`.

wwoodsTM replied:
I tried doing this but was not successful. llama_tokenize_internal seems to require a llama_vocab type for its first parameter and when I tried using model->vocab I got errors because model is incomplete for this scope. I wasn't sure how to fix this but if somebody can show me how to, I would be happy to make the change.

ngxson replied:
You can add another variant of `llama_tokenize_internal` that takes `model` instead of `vocab`, then expose it internally via `llama-impl.h`

```cpp
// wrapper function that takes "model" instead of "vocab", to be used internally
std::vector<llama_vocab::id> llama_tokenize_internal(
        const llama_model & model,
        std::string raw_text,
        bool add_special,
        bool parse_special) {
    return llama_tokenize_internal(model->vocab, raw_text, add_special, parse_special);
}
```

wwoodsTM replied:
@ngxson Thank you for the suggestion, but even with this approach I am still getting the same errors (`error: invalid use of incomplete type ‘const struct llama_model’`), unless I am doing something wrong?

I do have the two existing functions that are sitting in llama-impl.h for tokenizing and detokenizing that have been working but I have not had much luck pulling in other existing versions from other places in the codebase.

ngxson replied:
llama_model is only defined in llama.cpp. You must place the function body inside llama.cpp and expose the definition in llama-impl

wwoodsTM replied:
Indeed, that does work. Thank you! I just made a commit with this change.

ggerganov replied:
Hm, there was a much simpler solution. See how the `grammar` or `infill` samplers obtain an instance of `llama_vocab` and use that instead of propagating the `llama_model` to the samplers

ngxson replied:
Yeah that's right. On first look, I thought that `model` was used by DRY to access to various `llama_*` API. But turns out, only `llama_n_vocab` is currently using it, so passing the whole `model` is quite redundant.

@wwoodsTM What I can suggest now is to replace `llama_model * model` in `struct llama_sampler_dry` with more specific variables:
- `llama_vocab * vocab`
- `int n_ctx`
- `int n_vocab`

wwoodsTM commented:
> I am not sure if I understand correctly, but it is ok to store the strings in the sampler state as `std::string`, there is no cost to doing that, since you need to make a copy of the strings anyway to store them (can't just steal the pointers from the application). So hopefully changing this does not require converting the overlap code.

Ok, I have gone ahead and eliminated the C++ version of `llama_sampler_init_dry`. @slaren, as you suggested, I left the overlap function alone and basically did a `std::string` --> C-string --> `std::string` conversion process to smuggle the breakers through the C-interface.

wwoodsTM commented:
I have implemented all suggestions so far into the latest commit, including the use of `llama_tokenize_internal` as well as eliminating the support for JSON-within-JSON for passing sequence breakers. I also posted a message on SillyTavern's code-contribution channel and @Cohee1207 confirmed what @ngxson also pointed out here concerning TextGen WebUI being the origin of this format. @Cohee1207 expressed openness to a llama.cpp specific conversion so that ST would be compatible with this stricter implementation.

ggerganov commented on common/sampling.h:
```diff
@@ -36,7 +36,7 @@ struct common_sampler;
 // llama_sampler API overloads
 
-struct common_sampler * common_sampler_init(const struct llama_model * model, const struct common_sampler_params & params);
+struct common_sampler * common_sampler_init(const struct llama_model * model, const struct common_sampler_params & params, int32_t context_size);
```
Instead of adding this context size in so many places, better to use `llama_n_ctx_train(model)` when initializing the sampler. This value will always be larger or equal to the current context size, so it would be safe to use as a default value when DRY n_repeat is not specified. This way we avoid complicating the public interfaces with something very specific to the DRY sampler.

wwoodsTM replied:
@ggerganov I tried this running llama-server with `Mistral-Nemo-Instruct-2407-Q5_K_S.gguf` and CLI arguments ` -c 32000 -ngl 41 -b 2048 -t 6 -tb 12` and got a value that was higher by a factor of 32:
```
slot         init: id  0 | task -1 | new slot n_ctx_slot = 32000
... skipping some lines ...
Comparison between context_size values, old passed context size is: 32000
New calculated llama_n_ctx_train context size is: 1024000
```
Are we ok with it being that much higher?

wwoodsTM replied:
In looking further into this, I now see that this comes from hparams, so the code would allocate last_tokens and dry_repeat_count based on this theoretical max value for the model. Is that ok?

wwoodsTM replied:
The last time I coded in C++ prior to these PRs was in the 90s (wasn't kidding when I said I was rusty!), but I am probably too stuck in the old-school fashion of counting every kb of RAM used, so I just went ahead and implemented the llama_n_ctx_train(model) change as suggested. :sweat_smile: This restores the simplified common_sampler_init function signature.

ngxson reviewed (commented):
I've looked at the code of existing samplers to make this suggestion. Do you think this is good? @ggerganov @slaren

wwoodsTM commented on src/llama-sampling.cpp:
```diff
@@ -1660,6 +1660,408 @@ struct llama_sampler * llama_sampler_init_penalties(
+}
+
+// overloaded wrapper meant for test-sampling.cpp
+struct llama_sampler * llama_sampler_init_dry(const struct llama_model * model, int32_t context_size, float dry_multiplier, float dry_base, int32_t dry_allowed_length, int32_t dry_penalty_last_n, const std::vector<std::vector<llama_token>>& seq_breakers) {
```
Currently, I don't think we have access to a `llama_context` within `common_sampler_init` to be able to pass it to llama_sampler_init_dry unless I am missing something?

I had even considered forcing a `llama_context` reference through to common_sampler_init from `server.cpp` and `main.cpp`, but I went with `context_size` instead to keep it more focused as it seemed to be a less drastic change than what had already proved to be a bigger change than I would've wanted to begin with.

wwoodsTM replied:
I have gone ahead and followed @ggerganov's suggestion to use llama_n_ctx_train(model) for approximating the context size.

wwoodsTM commented:
@ggerganov I transitioned to using `llama_vocab` as suggested. This also allowed me to use "vocab.n_vocab" directly in the overlap function. I do not hold onto vocab in the sampler context as it is just needed for initialization and processing of the sequence breakers since I moved all of that into the main init function itself. 

I also moved the `llama_detokenize` function that takes `llama_vocab` to `llama-vocab.cpp` and restored `llama-impl.h` to its previous state.

wwoodsTM commented on src/llama-sampling.cpp:
```diff
@@ -1660,6 +1660,408 @@ struct llama_sampler * llama_sampler_init_penalties(
+// DRY
+
+struct llama_sampler_dry {
+    const   llama_model * model;
```
@ngxson I will leave it at your discretion if you want to resolve these conversations based on the current state of the code and if you are still looking for more feedback from @slaren and @ggerganov. My personal opinion is that by going with the llama_n_ctx_train(model) approach, the code is probably in a good state at this point and it also restores the function signature for common_sampler_init.

slaren reviewed (approved):
The sampler implementation looks good to me. I am not sure about the changes to the server, at least that will need to be cleaned up, but I will leave that review to @ngxson and @ggerganov.

wwoodsTM commented:
Sorry for all the commits and revisions that were necessary, and I appreciate everybody's patience and feedback through this process.

I just merged the latest changes from master, including the overhaul of test-sampling.cpp from several hours ago. I also made all of the recent suggested changes, including using llama_n_ctx_train(model) and the cleaner use of vocab.

Overall, this PR is in a much leaner and cleaner state because of everybody's valuable feedback, and thank you to everyone who offered it. Please let me know if I can do anything else, but I am definitely happy with how far it's come.

wwoodsTM commented:
> The sampler implementation looks good to me. I am not sure about the changes to the server, at least that will need to be cleaned up, but I will leave that review to @ngxson and @ggerganov.

Thank you for taking a look. 

As for the server code, I wanted to leave things commented just to be very clear where the changes were made until someone confidently said it was good to go. What I can say is that the movement of the sampling code in `server.cpp` definitely seemed to resolve that major bug that was reported before and that I discussed in the original post for the PR. And just to add some info, this bug was honestly worse than what was originally reported as it would not even pass the full prompt on further generations and would always be behind in handing tokens to the sampler system. As it stands, this process seems to now work smoothly and passes the complete prompt each time, but I am not sure about all use-cases or whether that somehow impacts something else inadvertently.

p-e-w commented on examples/main/README.md:
```diff
@@ -187,6 +187,27 @@ Use the `--no-penalize-nl` option to disable newline penalization when applying
+
+The `dry-base` option sets the base value for the exponential penalty calculation in DRY sampling. Higher values lead to more aggressive penalization of repetitions.
+
+The `dry-allowed-length` option determines the minimum length of repeated sequences that will be penalized. Repetitions shorter than or equal to this length are not penalized, allowing for natural repetitions of short phrases or common words.
```
The first sentence is a bit misleading, since it appears to imply that `dry-allowed-length` is "the minimum length of repeated sequences that will be penalized", when it is actually the *maximum* length of repeated sequences that will *not* be penalized.

wwoodsTM replied:
This change has also been implemented.

p-e-w commented on examples/main/README.md:
```diff
@@ -187,6 +187,27 @@ Use the `--no-penalize-nl` option to disable newline penalization when applying
+
+The `dry-allowed-length` option determines the minimum length of repeated sequences that will be penalized. Repetitions shorter than or equal to this length are not penalized, allowing for natural repetitions of short phrases or common words.
+
+The `dry-penalty-last-n` option controls how many recent tokens to consider when applying the DRY penalty. A value of -1 considers the entire context, while 0 disables this aspect of the penalty. Use a positive value to limit the consideration to a specific number of recent tokens.
```
The phrase "while 0 disables this aspect of the penalty" makes no sense. The penalty context is not an "aspect". If no context is considered, DRY does nothing.

wwoodsTM replied:
Took this out.

p-e-w commented on examples/main/README.md:
```diff
@@ -187,6 +187,27 @@ Use the `--no-penalize-nl` option to disable newline penalization when applying
+
+The `dry-penalty-last-n` option controls how many recent tokens to consider when applying the DRY penalty. A value of -1 considers the entire context, while 0 disables this aspect of the penalty. Use a positive value to limit the consideration to a specific number of recent tokens.
+
+DRY sampling works alongside traditional repetition penalties to provide more nuanced control over text generation, particularly for reducing long-range repetitions and maintaining global coherence.
```
FWIW, I actually designed DRY as a *replacement* for traditional repetition penalties, and I and many others use it as such. Running it alongside the standard repetition or frequency penalties is not something I recommend, as it brings back the grammar distortion effects associated with those samplers.

wwoodsTM replied:
I had made a mental note from your comment in the original PR [here](https://github.com/oobabooga/text-generation-webui/pull/5677#issuecomment-2053555111) about combining the two, but re-reading it now I can see that I obviously overgeneralized your comments too much. I took the language out.

p-e-w commented on examples/server/README.md:
```diff
@@ -357,6 +361,16 @@ node index.js
+
+    `dry_penalty_last_n`: How many tokens to scan for repetitions. Default: `-1`, where `0` is disabled and `-1` is context size.
+
+    `dry_sequence_breakers`: Specify an array of sequence breakers for DRY sampling. Can be provided as a JSON array of strings or as a JSON-encoded string representing an array of strings. Default: `["\n", ":", "\"", "*"]`
```
There was a discussion above mentioning a format incompatible with TGWUI that SillyTavern would need to accommodate, but the options described here (which include "a JSON-encoded string representing an array of strings") appear to be compatible with existing frontends. Has the proposed change not been implemented yet?

wwoodsTM replied:
The change was in fact made but this had not been updated yet to reflect that. Thank you for catching this.

p-e-w commented on examples/server/public/index-new.html:
```diff
@@ -833,15 +837,19 @@
           ${FloatField({ label: "Typical-P", title: "Activates local typical sampling, a method used to limit the prediction of tokens that are atypical in the current context. The parameter p controls the strength of this limitation. A value of 1.0 means that this function is deactivated.", max: 1.0, min: 0.0, name: "typical_p", step: 0.01, value: params.value.typical_p })}
           ${FloatField({ label: "XTC probability", title: "Sets the chance for token removal (checked once on sampler start)", max: 1.0, min: 0.0, name: "xtc_probability", step: 0.01, value: params.value.xtc_probability })}
           ${FloatField({ label: "XTC threshold", title: "Sets a minimum probability threshold for tokens to be removed", max: 0.5, min: 0.0, name: "xtc_threshold", step: 0.01, value: params.value.xtc_threshold })}
+          ${FloatField({ label: "DRY Penalty Multiplier", title: "Set the DRY repetition penalty multiplier. Default is 0.0, which is disabled.", max: 5.0, min: 0.0, name: "dry_multiplier", step: 0.01, value: params.value.dry_multiplier })}
```
```suggestion
          ${FloatField({ label: "DRY Penalty Multiplier", title: "Set the DRY repetition penalty multiplier. Default is 0.0, which disables DRY.", max: 5.0, min: 0.0, name: "dry_multiplier", step: 0.01, value: params.value.dry_multiplier })}
```

wwoodsTM replied:
Done.

p-e-w commented on examples/server/public/index-new.html:
```diff
@@ -833,15 +837,19 @@
           ${FloatField({ label: "XTC threshold", title: "Sets a minimum probability threshold for tokens to be removed", max: 0.5, min: 0.0, name: "xtc_threshold", step: 0.01, value: params.value.xtc_threshold })}
+          ${FloatField({ label: "DRY Penalty Multiplier", title: "Set the DRY repetition penalty multiplier. Default is 0.0, which is disabled.", max: 5.0, min: 0.0, name: "dry_multiplier", step: 0.01, value: params.value.dry_multiplier })}
+          ${FloatField({ label: "DRY Base", title: "Set the DRY repetition penalty base value. Default is 1.75", max: 3.0, min: 1.0, name: "dry_base", step: 0.01, value: params.value.dry_base })}
+          ${IntField({ label: "DRY Allowed Length", title: "Tokens that extend repetition beyond this receive exponentially increasing penalty. Default is 2", max: 10, min: 2, step: 1, name: "dry_allowed_length", value: params.value.dry_allowed_length })}
```
```suggestion
          ${IntField({ label: "DRY Allowed Length", title: "Tokens that extend repetition beyond this receive exponentially increasing penalty. Default is 2", max: 10, min: 1, step: 1, name: "dry_allowed_length", value: params.value.dry_allowed_length })}
```

I see no reason to disallow choosing `1` here. Even `0` should be possible in principle, though for that value DRY degenerates into something like the additive standard repetition penalty, and the implementation might not be ready for penalizing without sequence context. But there's nothing wrong with an allowed length of `1`, at least.

wwoodsTM replied:
Done.

p-e-w commented on examples/server/public/index-new.html:
```diff
@@ -833,15 +837,19 @@
+          ${FloatField({ label: "TFS-Z", title: "Activates tail-free sampling, a method used to limit the prediction of tokens that are too frequent. The parameter z controls the strength of this limitation. A value of 1.0 means that this function is deactivated.", max: 1.0, min: 0.0, name: "tfs_z", step: 0.01, value: params.value.tfs_z })}
           ${IntField({ label: "Min Keep", title: "If greater than 0, samplers are forced to return N possible tokens at minimum. Default is 0", max: 10, min: 0, name: "min_keep", value: params.value.min_keep })}
-        </fieldset>
+          </fieldset>
```
```suggestion
        </fieldset>
```

Matches indentation of the opening tag.

wwoodsTM replied:
Fixed.

p-e-w commented on examples/server/public/style.css:
?

wwoodsTM replied:
I am not sure why GitHub is showing this as empty in this view If you click the three dots and view the file, or view the raw file, it seems to show it that way.

wwoodsTM commented on common/arg.cpp:
```diff
@@ -997,6 +997,34 @@ common_params_context common_params_parser_init(common_params & params, llama_ex
+    ).set_sparam());
+    add_opt(common_arg(
+        {"--dry-penalty-last-n"}, "N",
+        string_format("Set DRY penalty for the last n tokens (default: %d, 0 = disable, -1 = context size)", params.sparams.dry_penalty_last_n),
```
My thought was that to maintain consistency with the other llama.cpp options such as `repeat-last-n`, `n-keep`,  `n-predict`, etc. where -1 means all or infinity, that this would probably be the best compromise.

wwoodsTM commented on common/arg.cpp:
```diff
@@ -997,6 +997,34 @@ common_params_context common_params_parser_init(common_params & params, llama_ex
+        [](common_params & params, int value) {
+            params.sparams.dry_penalty_last_n = value;
+        }
+    ).set_sparam());
```
I like your idea of using the reusable singular "--dry-sequence-breaker" CLI argument and went ahead and implemented that. It took some finagling of dealing with clearing out the defaults when one is used and having a "none" option, but I think I struck a decent balance with it.

wwoodsTM commented on common/common.cpp:
```diff
@@ -2019,6 +2019,10 @@ void yaml_dump_non_result_info(FILE * stream, const common_params & params, cons
     fprintf(stream, "chunks: %d # default: -1 (unlimited)\n", params.n_chunks);
     fprintf(stream, "color: %s # default: false\n", params.use_color ? "true" : "false");
     fprintf(stream, "ctx_size: %d # default: 512\n", params.n_ctx);
+    fprintf(stream, "dry_allowed_length: %d # default: 2\n", sparams.dry_allowed_length);
```
Yeah, it struck me as a bit odd the first time I saw the ordering too.

wwoodsTM commented on common/common.cpp:
```diff
@@ -2019,6 +2019,10 @@ void yaml_dump_non_result_info(FILE * stream, const common_params & params, cons
+    fprintf(stream, "dry_allowed_length: %d # default: 2\n", sparams.dry_allowed_length);
+    fprintf(stream, "dry_base: %.2f # default: 1.75\n", sparams.dry_base);
+    fprintf(stream, "dry_multiplier: %.1f # default: 0.0\n", sparams.dry_multiplier);
+    fprintf(stream, "dry_penalty_last_n: %d # default: 0 (0 = disable, -1 = context size)\n", sparams.dry_penalty_last_n);
```
Agreed, and you were right that this was a typo. Clearly, the documentation was too rushed when I did this originally and I thank you for catching these issues.

wwoodsTM commented on common/common.h:
```diff
@@ -104,32 +105,39 @@ enum dimre_method {
+    float   dry_multiplier     = 0.0f;  // 0.0 = disabled;      DRY repetition penalty for tokens extending repetition:
+    float   dry_base           = 1.75f; // 0.0 = disabled;      multiplier * base ^ (length of sequence before token - allowed length)
+    int32_t dry_allowed_length = 2;     // tokens extending repetitions beyond this receive penalty
+    int32_t dry_penalty_last_n = -1;    // how many tokens to scan for repetitions (0 = disable penalty, -1 = context size)
```
Correct, it was a typo.

wwoodsTM commented on common/common.h:
```diff
@@ -104,32 +105,39 @@ enum dimre_method {
+    float   penalty_freq       = 0.00f; // 0.0 = disabled
+    float   penalty_present    = 0.00f; // 0.0 = disabled
+    float   dry_multiplier     = 0.0f;  // 0.0 = disabled;      DRY repetition penalty for tokens extending repetition:
+    float   dry_base           = 1.75f; // 0.0 = disabled;      multiplier * base ^ (length of sequence before token - allowed length)
```
Thank you for the suggestion, I went ahead and implemented this.

wwoodsTM commented on examples/main/README.md:
```diff
@@ -187,6 +187,27 @@ Use the `--no-penalize-nl` option to disable newline penalization when applying
 
+### DRY Repetition Penalty
+
+DRY (Don't Repeat Yourself) sampling is an effective technique for reducing repetition in generated text even across long contexts by penalizing tokens based on their recent usage patterns.
```
Agreed. The link has been added.

ngxson reviewed (changes requested):
To make the review less complicated, let's try not to introduce changes unrelated to the current PR.

wwoodsTM commented:
@p-e-w Thank you for all your many good catches there and your suggestion for the `--dry-sequence-breaker` singular CLI argument which I have implemented.

wwoodsTM commented on examples/server/server.cpp:
```diff
@@ -164,7 +164,7 @@ struct server_slot {
     int32_t n_prompt_tokens_processed = 0;
 
-    json prompt; // can be either a string, array of strings or array of token ids
+    json prompt; // can be either a string, array of strings, array of token ids, or mixed array of strings and token ids
```
If it is an array of strings, the code treats it properly as a multi-task. It is only in the case of a mixture of tokens and strings that it is treated as a singleton mixture. To me, this seemed like a good compromise.

I understand it may be out of the scope of this particular PR, but to be honest, this issue is pretty near and dear to me, and is the main reason I switched to using llama-server to begin with. Could I separate out the code into another PR or do you feel it has fundamental problems?

ngxson replied:
> If it is an array of strings, the code treats it properly as a multi-task. It is only in the case of a mixture of tokens and strings that it is treated as a singleton mixture. To me, this seemed like a good compromise.

Ok I wasn't aware that we have ability to used mixed type. Before, something like `[12, 34, "abc", 56, 78],` won't work.

This must be refactor anyways, because the code is getting spaghetti after all of this "added features". What I suggest is that you can revert these change and I'll make a PR to refactor the task splitting and tokenization. Ideally, `slot.prompt` should be an array of tokens, it should not hold a string anymore.

wwoodsTM replied:
Understood. Thank you for being open to addressing this in the refactor. It is an extremely helpful feature for front-end software when you need to get into the nitty-gritty of prompting outside of an external chat-template and directly apply your own templates or experiment with different prompt styles for models that accept multiple formats, all directly using the `/completion` endpoint.

With the ability to mix tokens and strings in one prompt I can rapidly test different models and prompting styles without having to necessarily handle a bunch of different tokenizing setups internally in my own code or do a bunch of separate fetches to the server to tokenize.

But yes, I will remove this specific code in the next commit, thank you.

ngxson replied:
I'm trying to refactor this part in next few days, will push a PR when it's ready

wwoodsTM commented on examples/server/server.cpp:
```diff
@@ -2163,16 +2213,19 @@ struct server_context {
-                                }
+                                // for (int i = 0; i < slot.n_past; ++i) {
+                                //     common_sampler_accept(slot.smpl, slot.cache_tokens[i], false);
+                                // }
```
I added a line to print `prompt_tokens.size()` in this part:

```
                    // entire prompt has been processed
                    if (slot.n_past == slot.n_prompt_tokens) {
                        slot.state = SLOT_STATE_DONE_PROMPT;

                        GGML_ASSERT(batch.n_tokens > 0);

                        common_sampler_reset(slot.smpl);

                        // Process all prompt tokens through sampler system
                        for (int i = 0; i < slot.n_prompt_tokens; ++i) {
                            common_sampler_accept(slot.smpl, prompt_tokens[i], false);
                        }

                        // extract the logits only for the last token
                        batch.logits[batch.n_tokens - 1] = true;

                        slot.n_decoded = 0;
                        slot.i_batch   = batch.n_tokens - 1;

                        SLT_INF(slot, "size of prompt_tokens array = %ld\n", prompt_tokens.size());
                        SLT_INF(slot, "prompt done, n_past = %d, n_tokens = %d\n", slot.n_past, batch.n_tokens);
                    }
```
And got this output on my second generation with `'cache_prompt': true` using a decently large context:

```
slot launch_slot_: id  0 | task 215 | processing task
slot update_slots: id  0 | task 215 | tokenizing prompt, len = 233
slot update_slots: id  0 | task 215 | prompt tokenized, n_ctx_slot = 32000, n_keep = 0, n_prompt_tokens = 9483
slot update_slots: id  0 | task 215 | kv cache rm [8816, end)
slot update_slots: id  0 | task 215 | prompt processing progress, n_past = 9483, n_tokens = 667, progress = 0.070336
slot update_slots: id  0 | task 215 | size of prompt_tokens array = 9483
slot update_slots: id  0 | task 215 | prompt done, n_past = 9483, n_tokens = 667

---- DRY sampling debug info:
  Applied penalty 0.8000 to token 1395 ( is) (repeat length 2)
  Applied penalty 0.8000 to token 1681 ('s) (repeat length 2)
  Applied penalty 0.8000 to token 10637 ( looks) (repeat length 2)
  Applied penalty 0.8000 to token 45493 ( beams) (repeat length 2)
  Applied penalty 0.8000 to token 86372 ( nods) (repeat length 2)
  Applied penalty 0.8000 to token 92156 ( shakes) (repeat length 2)
Total DRY penalties applied: 6
DRY Ring Buffer Content (showing 9485 out of 9485 tokens):
```

Both `prompt_tokens.size()` and my ring buffer `last_tokens` are showing the full size of the context (with `last_tokens` being larger by 2 on the last line because my debugging doesn't output until DRY is actually triggered).

I'm not sure if this is a bug that this works, but it seems to be working? I also have a full print out in my debugging code that shows the entirety of the prompt being present, token by token, for all 9483 tokens.

wwoodsTM commented on examples/server/server.cpp:
```diff
@@ -2288,6 +2341,13 @@ struct server_context {
+                        for (int i = 0; i < slot.n_prompt_tokens; ++i) {
+                            common_sampler_accept(slot.smpl, prompt_tokens[i], false);
+                        }
+
```
@ngxson @ggerganov Can I keep this code in the PR as a temporary bandaid since it seems to potentially function a bit better than the current state at present, and then after the PR merges this could be fixed more properly?

ggerganov replied:
@ngxson and @wwoodsTM This indeed looks like a bug and wasn't the intended way.

The changes here appear to be mostly correct, except the sampler will now accept the tokens even for embedding and reranking tasks, which is probably completely fine.

> Removing this will break the cached prompt. These tokens are not in prompt_tokens

I think the `prompt_tokens` should contain the full context, including the cached tokens, so this should be OK.

Like @ngxson mentioned earlier, I agree that the sampler changes in the server code would be better to be reviewed in a separate PR. @wwoodsTM Would you mind to open a new PR just for those changes alone - I think we can merge them fast and then you can rebase the DRY implementation on top of them?

wwoodsTM replied:
@ggerganov Makes total sense to me, thanks.

I just created #10019 for this, as requested.

wwoodsTM commented:
I have removed the mixed arrays due to that issue being out of scope of the PR, as requested by @ngxson. 

To my knowledge, the only pending thing at this point is the question above concerning whether I am OK keeping my bandaid fix in the PR as a bridge to a more proper solution.

wwoodsTM commented:
@ggerganov @ngxson The separate PR #10019 for the server sampling code is up and just passed all checks. I removed all of the commented-out code from it and it left it with just the precise removals and additions needed. I also went ahead and resolved all the comments up above to reflect the current status. Thank you.

wwoodsTM committed:
sampling : add DRY sampler (post-refactor) #9702

Co-authored-by: l3utterfly <gc.pthzfoldr@gmail.com>
Co-authored-by: pi6am <34464159+pi6am@users.noreply.github.com>

* Implementation of DRY Sampling (post-sampling-refactor)

* Better merge of latest changes to master in server.cpp

* Added co-authors, testing, fix for consistent prompt sampling

* Misc. cleanup

* Remove debug define

* Fixing strict compile errors, cleaning up

* Removed unnecessary include

* Removed debugging code, moved (de)tokenizing functions

* Bug fixes to z-alg implementation

* small fix for string_format

* restoring Koboldcpp comments, adding attribution in llama.h and llama-sampling.cpp

* fixes, adjustments from code review

* Removing sequence breaker functions

* Addressed warnings from pedantic compile

* Trying alternative ways to pull in external tokenize/detoneize functions

* Removed C++ version of llama_sampler_init_dry

* Implemented testing code

* Switched to using llama_tokenize_internal

* Removed support for JSON-within-JSON for seq breakers

* Removed old llama_sampler_init_dry declaration

* Trying ggerganov's suggestion with llama_n_ctx_train

* Using vocab instead of model

* Restored common_sampler_init signature, using llama_n_ctx_train for context, updated testing code

* Minor cleanup

* Added --dry-sequence-breaker arg for single breaker, fixed docs and default values

* Removed mixed arrays for prompt

wwoodsTM committed:
DRY: Trying to fix coauthors, removed unneeded line

Co-authored-by: l3utterfly <gc.pthzfoldr@gmail.com>
Co-authored-by: pi6am <34464159+pi6am@users.noreply.github.com>

wwoodsTM commented:
Not sure what I did wrong with the "co-authored" part in the first force-pushes as GitHub didn't seem to want to recognize that; maybe it deals with force-pushes differently in that respect? Anyway, I removed an extra empty line that I probably should've removed originally anyway and did a normal commit to make sure that @l3utterfly and @pi6am were restored as co-authors on this PR.

wwoodsTM commented:
@ggerganov @slaren @ngxson 

Is it possible my quick force-pushes and commit all in quick succession threw off the tests and that's why we're getting failures?

slaren commented:
Some of the builds have warnings as errors, it fails to build here:
>examples/server/server.cpp:908:18: error: unused variable 'dry_sequence_breakers' [-Werror,-Wunused-variable]

wwoodsTM committed:
DRY: Fixed redundant code

wwoodsTM commented:
> Some of the builds have warnings as errors, it fails to build here:
> 
> > examples/server/server.cpp:908:18: error: unused variable 'dry_sequence_breakers' [-Werror,-Wunused-variable]

Thank you! Just recommitted.

I thought I had my warnings and pedantic settings all cranked to the max but somehow that one didn't show for me.

wwoodsTM commented:
I am able to reproduce the crash on my own machine coming from running this:
`./bin/llama-cli -m stories260K.gguf -p "One day, Lily met a Shoggoth" -n 500 -c 256` 

I am now trying to track down the issue.

wwoodsTM committed:
DRY: Fixed crash issue due to DRY being in chain but uninitialized

wwoodsTM commented:
@ggerganov @ngxson I resolved the crash issue, which was due to DRY being in the default sampler chain but not being initialized when disabled. Now DRY will properly initialize with empty data structures even when disabled.

ngxson reviewed (approved):
Yup thanks, the code looks much cleaner now.

wwoodsTM commented:
> Yup thanks, the code looks much cleaner

Thank you, it was greatly helped by your guidance and that of everyone else who has offered feedback.

@ggerganov Do you feel this is ready to merge?

ggerganov commented on src/llama-vocab.cpp:
```diff
@@ -1966,3 +1966,19 @@ int32_t llama_detokenize_impl(
 
     return total <= text_len_max ? total : -total;
 }
+
+std::string llama_detokenize(const struct llama_vocab & vocab, const std::vector<llama_token> & tokens, bool special) {
+    std::string text;
+    text.resize(std::max(text.capacity(), tokens.size()));
+    int32_t n_chars = llama_detokenize_impl(vocab, tokens.data(), (int32_t)tokens.size(), &text[0], (int32_t)text.size(), false, special);
+    if (n_chars < 0) {
+        text.resize(-n_chars);
+        n_chars = llama_detokenize_impl(vocab, tokens.data(), (int32_t)tokens.size(), &text[0], (int32_t)text.size(), false, special);
+        GGML_ASSERT(n_chars <= (int32_t)text.size());  // whitespace trimming is performed after per-token detokenization
+    }
+
+    text.resize(n_chars);
+
+    // NOTE: the original tokenizer decodes bytes after collecting the pieces.
+    return text;
+}
```
It might be better to not introduce this function and instead maintain a string in the sampler context that is resized as needed (see the infill sampler for a similar idea). The main goal would be to avoid memory allocations that would currently happen here for each token. Can be done in follow-up PR

wwoodsTM replied:
In this case, the function is only used for initialization with the sequence breakers so the impact should hopefully not be too great. In the case of Nemo, this process resulst in the initial default 4 sequence breakers transforming into over 2000 sequence breakers because @pi6am’s overlap code includes any tokens that overlap with the 4 sequence breakers. But this process only happens on initialization and the processed sequence breakers are then kept in token form from then on.

ggerganov replied:
Got it - I missed that this is used only during init.

ggerganov reviewed (approved):
@wwoodsTM Good job!

ExtReMLapin commented:
damn, what a ride

thanks @wwoodsTM for your work !

wwoodsTM commented:
> @wwoodsTM Good job!

@ggerganov Thank you, I learned a lot throughout this process and have appreciated your feedback throughout.

I am really excited to see this merge, let me know if I need to do anything else on my end for that step. I also hope if there is a squash that the co-authored parts survive better than when I did my force-pushes.

MaggotHATE commented:
Sorry for barging in, but what's the reason for DRY being a sampler in queue and not a sampler _before_ the queue, like penalties? It used to go right after penalties and before the queue in the first PR, and I'm not sure there is a reason to have DRY anywhere except at the beginning of the queue.

wwoodsTM commented:
> Sorry for barging in, but what's the reason for DRY being a sampler in queue and not a sampler _before_ the queue, like penalties? It used to go right after penalties and before the queue in the first PR, and I'm not sure there is a reason to have DRY anywhere except at the beginning of the queue.

@MaggotHATE You are right that I originally implemented it as occurring between traditional penalties and the sampler chain. The change to add it into the chain stemmed from @compilade's request in this [review comment](https://github.com/ggerganov/llama.cpp/pull/9702#discussion_r1792270167). I also like you wonder whether there is much applicability in other places, but I felt that I could address his request and expand flexibility a bit without hurting anything if I put it first in the chain.

MaggotHATE commented:
> from @compilade's request in this [review comment](https://github.com/ggerganov/llama.cpp/pull/9702#discussion_r1792270167).

Oh, sorry, looks like I've missed this comment. If it is by request, then it's ok. I'm still not sure where else it might be used, but more flexibility is definitely for better.
